### PR TITLE
Fix StorageMerge with aliases and where

### LINKED ISSUE #28802: `Merge` table doesn't work with aliases and `WHERE`
**Does it reproduce on recent release?**

Reproduces on master.

**How to reproduce**

```sql
DROP TABLE IF EXISTS tt1;
DROP TABLE IF EXISTS tt2;

CREATE TABLE tt1 (a UInt32, b UInt32 ALIAS 0) ENGINE = Memory;
CREATE TABLE tt2 (a UInt32, b UInt32 ALIAS 0) ENGINE = Memory;
CREATE TABLE tt_m (a UInt32, b UInt32) ENGINE = Merge(currentDatabase(), 'tt1|tt2');

INSERT INTO tt1 VALUES (1);
INSERT INTO tt1 VALUES (2);
```

```sql
SELECT * FROM tt_m;

┌─a─┬─b─┐
│ 1 │ 0 │
└───┴───┘
┌─a─┬─b─┐
│ 2 │ 0 │
└───┴───┘
```

```sql
SELECT * FROM tt_m WHERE b != 0;

Received exception from server (version 21.11.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column b in block. There are only columns: a. (NOT_FOUND_COLUMN_IN_BLOCK)
```
```sql
SELECT * FROM tt_m WHERE a = 1;

Received exception from server (version 21.11.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column b in block. There are only columns: a. (NOT_FOUND_COLUMN_IN_BLOCK)
```

```sql
SELECT * FROM tt_m WHERE 0;

Received exception from server (version 21.11.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column b in block. There are only columns: a. (NOT_FOUND_COLUMN_IN_BLOCK)
```

```sql
SELECT * FROM tt_m WHERE 1;

┌─a─┬─b─┐
│ 2 │ 0 │
└───┴───┘
┌─a─┬─b─┐
│ 1 │ 0 │
└───┴───┘
```

### PR DESCRIPTION
**Pull request by kssenii**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix StorageMerge with aliases and where (it did not work before at all). Closes https://github.com/ClickHouse/ClickHouse/issues/28802.

### CODE REVIEW

File: src/Storages/StorageMerge.cpp [hunk 1]
```diff
@@ -285,9 +285,16 @@ Pipe StorageMerge::read(
 
         if (processed_stage == QueryProcessingStage::FetchColumns && !storage_columns.getAliases().empty())
         {
+            ASTPtr where_expression;
+            if (query_info.query->as<ASTSelectQuery>()->where())
+                where_expression = query_info.query->as<ASTSelectQuery>()->where()->clone();
+
```
Comment (CurtizJ) [inline src/Storages/StorageMerge.cpp:291]: So, the problem is, that aliases are replaced by theirs expression during query analysis, right? Maybe we have the same problem with other expressions?

Comment (CurtizJ) [reply src/Storages/StorageMerge.cpp:291]: Looks, like we have. It still doesn't work if we query only alias column.

```sql
DROP TABLE IF EXISTS tt1;
DROP TABLE IF EXISTS tt2;
DROP TABLE IF EXISTS tt3;
DROP TABLE IF EXISTS tt_m;

CREATE TABLE tt1 (a UInt32, b UInt32 ALIAS a, c UInt32) ENGINE = Memory;
CREATE TABLE tt2 (a UInt32, b UInt32 ALIAS a * 2, c UInt32) ENGINE = Memory;
CREATE TABLE tt3 (a UInt32, b UInt32 ALIAS 0, c UInt32) ENGINE = Memory;
CREATE TABLE tt_m (a UInt32, b UInt32, c UInt32) ENGINE = Merge(currentDatabase(), 'tt1|tt2|tt3');

INSERT INTO tt1 (a, c) VALUES (1, 1);
INSERT INTO tt2 (a, c) VALUES (2, 2);
INSERT INTO tt3 (a, c) VALUES (3, 3);
```

Comment (CurtizJ) [reply src/Storages/StorageMerge.cpp:291]:
```sql
SELECT b  FROM tt_m WHERE b = 1

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b, count()  FROM tt_m GROUP BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b FROM tt_m ORDER BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b FROM tt_m LIMIT 1 BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
````

Comment (CurtizJ) [reply src/Storages/StorageMerge.cpp:291]: If we add one more column (which doesn't participate in alias expression) some queries starts working, but some not.

```sql
SELECT c FROM tt_m WHERE b = 1

┌─c─┐
│ 1 │
└───┘
```

```sql
SELECT sum(c) FROM tt_m group by b

┌─sum(c)─┐
│      3 │
│      2 │
│      1 │
└────────┘
```

```sql
SELECT c FROM tt_m ORDER BY b

Received exception from server (version 21.12.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column a in block. There are only columns: b, c. (NOT_FOUND_COLUMN_IN_BLOCK)
```

```sql
SELECT c FROM tt_m LIMIT 1 BY b

Received exception from server (version 21.12.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column a in block. There are only columns: b, c. (NOT_FOUND_COLUMN_IN_BLOCK)
```

Comment (kssenii) [reply src/Storages/StorageMerge.cpp]: Fixed

File: src/Storages/StorageMerge.cpp [hunk 2]
```diff
@@ -364,19 +395,13 @@ Pipe StorageMerge::createSources(
...
-    SelectQueryInfo modified_query_info = query_info;
-    modified_query_info.query = query_info.query->clone();
-
     /// Original query could contain JOIN but we need only the first joined table and its columns.
     auto & modified_select = modified_query_info.query->as<ASTSelectQuery &>();
 
-    TreeRewriterResult new_analyzer_res = *query_info.syntax_analyzer_result;
+    TreeRewriterResult new_analyzer_res = *modified_query_info.syntax_analyzer_result;
     removeJoin(modified_select, new_analyzer_res, modified_context);
     modified_query_info.syntax_analyzer_result = std::make_shared<TreeRewriterResult>(std::move(new_analyzer_res));
```
Comment (CurtizJ) [inline src/Storages/StorageMerge.cpp:403]: Looks like it's unneeded, because it's already done in `getModifiedQueryInfo`.

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**